python/src/iceberg/utils/schema_conversion.py

```diff
@@ -0,0 +1,451 @@
+
+        Returns:
+            A tuple containing the type and nullability
+
+        Raises:
+            TypeError: In the case non-optional union types are encountered
+        """
+        avro_types: Dict | List
+        if isinstance(type_union, str):
+            # It is a primitive and required
+            return type_union, False
+        elif isinstance(type_union, dict):
+            # It is a context and required
+            return type_union, False
+        else:
+            avro_types = type_union
+
+        is_optional = "null" in avro_types
+
+        if len(avro_types) > 2:
+            raise TypeError("Non-optional types aren't part of the Iceberg specification")
+
+        # Filter the null value and return the type
+        return list(filter(lambda t: t != "null", avro_types))[0], is_optional
+
+    def _resolve_inner_type(self, avro_type: Dict[str, Any], inner_field_name: str, id_field: str) -> Tuple[IcebergType, bool]:
+        """
+        Removes any unions inside an array or map, and appends the field-id downstream (if required)
+
+        Args:
+            avro_type: The parent Avro type
+            inner_field_name: For example 'items' in case of an array
+            id_field: Contains the element or key/value id, for example element-id, to be copied to the field-id
+
+        Returns:
+            The inner type and the nullability
+        """
+        plain_type, element_is_optional = self._resolve_union(avro_type[inner_field_name])
+        if isinstance(plain_type, dict):
+            # We need the element-id downstream
+            plain_type["field-id"] = avro_type[id_field]
+        inner_field = self._convert_schema(plain_type)
+        return inner_field, element_is_optional
+
+    def _convert_schema(self, avro_type: str | Dict[str, Any]) -> IcebergType:
+        """
+        Resolves the Avro type
+
+        Args:
+            avro_type: The Avro type, can be simple or complex
+
+        Returns:
+            The equivalent IcebergType
+
+        Raises:
+            ValueError: When there are unknown types
+        """
+        if isinstance(avro_type, str):
+            return PRIMITIVE_FIELD_TYPE_MAPPING[avro_type]
+        elif isinstance(avro_type, dict):
+            if "logicalType" in avro_type:
+                return self._convert_logical_type(avro_type)
+            else:
+                type_identifier = avro_type["type"]
+                if type_identifier == "record":
+                    return self._convert_record_type(avro_type)
+                elif type_identifier == "array":
+                    return self._convert_array_type(avro_type)
+                elif type_identifier == "map":
+                    return self._convert_map_type(avro_type)
+                elif type_identifier == "fixed":
+                    return self._convert_fixed_type(avro_type)
+                else:
+                    raise ValueError(f"Unknown type: {avro_type}")
```

Comment:
It's uncommon, but it's valid for a primitive to be wrapped in a dict, like `{"type": "array", "items": {"type": "double", "additional": "property"}}`. That should be handled here. I think you can check whether `type_identifier` is in the primitive map.

Reply:
Good one, haven't seen that myself. Added a test and logic for handling this